src/renderer/components/ReactFlowBox.tsx

```diff
@@ -216,8 +286,200 @@ export const ReactFlowBox = memo(({ wrapperRef, nodeTypes, edgeTypes }: ReactFlo
         return [displayNodes, displayEdges, isSnapToGrid && snapToGridAmount];
     }, [nodes, edges, isSnapToGrid, snapToGridAmount]);
 
+    // Node-on-edge collision detection
+    const performNodeOnEdgeCollisionDetection = useCallback(
+        (node: Node<NodeData>, mousePosition: XYPosition) => {
+            // First, we need to make sure this node is an orphan. We can do a find so it stops early
+            const hasConnectedEdge = !!edges.find(
+                (e) => e.source === node.id || e.target === node.id
+            );
+            if (hasConnectedEdge) {
+                return;
+            }
+
+            const nodePos: Point = { x: node.position.x || 0, y: node.position.y || 0 };
+            const nodeBB = AABB.fromPoints(nodePos, {
+                x: nodePos.x + (node.width || 0),
+                y: nodePos.y + (node.height || 0),
+            });
+
+            const fn = functionDefinitions.get(node.data.schemaId);
+            if (!fn) {
+                return;
+            }
+
+            // Finds the first edge that intersects with the node
+            type CandidateEdge = Edge<EdgeData> & {
+                data: Required<EdgeData>;
+                sourceHandle: string;
+                targetHandle: string;
+            };
+            const intersectingEdges = edges
+                .filter((e): e is CandidateEdge => {
+                    // if one value is set, all are
+                    return Boolean(
+                        e.data?.sourceX !== undefined && e.sourceHandle && e.targetHandle
+                    );
+                })
+                .flatMap((e) => {
+                    const sourceP: Point = { x: e.data.sourceX, y: e.data.sourceY };
+                    const targetP: Point = { x: e.data.targetX, y: e.data.targetY };
+
+                    // check node and edge bounding boxes
+                    const edgeBB = AABB.fromPoints(sourceP, targetP);
+                    if (!nodeBB.intersects(edgeBB)) {
+                        return EMPTY_ARRAY;
+                    }
+
+                    // Check if the node has valid connections it can make
+                    // If it doesn't, we don't need to bother checking collision
+                    const { outputId } = parseSourceHandle(e.sourceHandle);
+                    const edgeType = typeState.functions.get(e.source)?.outputs.get(outputId);
+                    if (!edgeType) {
+                        return EMPTY_ARRAY;
+                    }
+                    const { inputId } = parseTargetHandle(e.targetHandle);
+                    const targetNode = reactFlowInstance.getNode(e.target);
+                    if (!targetNode) {
+                        return EMPTY_ARRAY;
+                    }
+                    const targetEdgeType = functionDefinitions
```

Comment:
Why didn't you use typeState?

Reply:
I don't want to check based on the current type of the input, I want to check what the type is without any connections. Otherwise, we'd have the same problem as in the screenshot you sent (we talked about this)

Unless there's something in the typestate that can do this that I'm unaware of?

Reply:
TypeState functions have a reference to the type definition, so you don't need to call `reactFlowInstance.getNode`. You can do `typeState.functions.get(e.target)?.definition.inputDefaults.get(inputId)`.

Reply:
Oh nice, I'll do that then